### PR TITLE
Add implicit cast for 'sum' and 'count'

### PR DESCRIPTION
**Pull request by jordalgo**
This utilizes 'map_lookup_percpu_elem' to iterate over the per-cpu maps that 'sum' and 'count' use to aggregate the values for each cpu so these maps can be used in kernel space for conditionals and arithmetic.

Additionally, add explicit casting for printing or storing just the value for these types of maps.

Issue: https://github.com/bpftrace/bpftrace/issues/3126

Note: We can do the same to support to 'avg', 'max', 'min' but those require the same user-space logic written in llvm ir.

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

File: src/ast/passes/semantic_analyser.cpp
Comment (danobi) [inline src/ast/passes/semantic_analyser.cpp:1794]: I'm wondering if we can be more generic here. For example, this doesn't quite fix Jesper's issue where he wanted to printf() the count() value. 

Perhaps we can allow casts to int64/int32/etc from Type::count? I kinda like the idea of explicit casts cuz it'll at least hint we're gonna do something magical. Eg.

```
@ = count();
printf("%d\n", (int64)@);
```

WDYT?

**Comment by jordalgo** [comment]
I'll update the changelog and adoc once the code is approved.

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [reply src/ast/passes/semantic_analyser.cpp:1794]: Yeah, I like explicit casts - let me look into supporting that.

**Comment by jordalgo** [comment]
Note: Also going to add a codegen test for this

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [reply src/ast/passes/semantic_analyser.cpp:1794]: Added this in the latest version.

**Comment by jordalgo** [comment]
Update: I added a bit of a hack to make this work for AOT. Basically, when in AOT mode, just set an upper max on the number of cpus to iterate over (1024 - which is still below some crazy new machines but the verifier won't freak out) and exit the loop early if we fail to look up the value for that cpu. I tested this locally and it worked as expected.

File: src/ast/irbuilderbpf.cpp
Comment (danobi) [inline src/ast/irbuilderbpf.cpp:551]: I forget if we discussed this in person, did it not work to pass the nr_cpu from userspace and then mask (`&=`) the count?

If it didn't and we have to use this method, since 1024 is a fixed value, can bpftrace check at runtime if there's > 1024 CPUs and warn about it? This is a potential for incorrect results & and we should be really careful in these situations. Should be possible in a shared codepath between AOT and dynamic mode.

File: src/ast/passes/codegen_llvm.cpp
Comment (danobi) [inline src/ast/passes/codegen_llvm.cpp:1427]: Can we add a helper method (maybe in types.h?) that groups cast-able map types? It'll help make things more clear across semantic analysis && codegen

**Review by danobi — COMMENTED** [review]
could we add some semantic analysis test cases to check that types are coerced or casted where we expect?

File: tests/codegen/per_cpu_map_cast.cpp
Comment (danobi) [inline tests/codegen/per_cpu_map_cast.cpp:1]: Could we add a codegen test for `BEGIN { @ = count(); print(@) }` ? This would help ensure that we don't loop unless we actually need to.

File: src/ast/irbuilderbpf.cpp
Comment (danobi) [inline src/ast/irbuilderbpf.cpp:356]: I think we should add a `CreateHelperErrorCond` call here too

File: src/ast/irbuilderbpf.h
Comment (danobi) [inline src/ast/irbuilderbpf.h:98]: Is this unused? I deleted it from my local checkout and things still compile && work

File: src/ast/irbuilderbpf.cpp
Comment (danobi) [inline src/ast/irbuilderbpf.cpp:626]: Should `getInt32(0)` be -ENOENT ?

File: src/ast/irbuilderbpf.h
Comment (jordalgo) [reply src/ast/irbuilderbpf.h:98]: Good catch. I think this was leftover from a previous revision

File: src/ast/irbuilderbpf.cpp
Comment (jordalgo) [reply src/ast/irbuilderbpf.cpp:626]: This mirrors what we do for `CreateMapLookupElem` as the return value is null if the map element doesn't exist.

File: src/ast/irbuilderbpf.cpp
Comment (jordalgo) [reply src/ast/irbuilderbpf.cpp:356]: We don't do this for `createMapLookup` because the return value is the map value or null not an error code.

File: src/ast/irbuilderbpf.cpp
Comment (jordalgo) [reply src/ast/irbuilderbpf.cpp:551]: > I forget if we discussed this in person, did it not work to pass the nr_cpu from userspace and then mask (&=) the count?

I've tried other limiting techniques in other PRs and it seems like the verifier is unhappy if it doesn't know the value in advance and the type can exceed the max number of loops. You can normally get around this using libbpf by having a mutable volatile with an initial value and modifying it using the skeleton in userspace but we don't have that yet. @ajor and I discussed the option of doing this by modifying the rodata section but decided just to punt on that since AOT isn't officially supported yet anyway. 

I will add a warning if we detect the number of cpus to be greater than 1024 in userspace.

Comment (jordalgo) [reply src/ast/irbuilderbpf.cpp:551]: I also have plans to add proper loops using open coded iterators which should also fix this issue.

File: src/main.cpp
Comment (danobi) [inline src/main.cpp:886]: I think this warning is only relevant if the _runtime_ host has >1024 CPUs. So if build host had >1024 CPUs but runtime had <1024, it'd still be ok. So I think this should go in (or near) `aot_main.cpp`. 

Having this hard coded message is probably OK if you plan on removing it soonish. Was thinking if it makes sense to analyze RequiredResources in aot.cpp::load_required_resources. It'd be more accurate -- we could check if count() or sum() is even being used. IMO it'd be a little cleaner to have it there. Up to you.

Comment (jordalgo) [reply src/main.cpp:886]: > So I think this should go in (or near) aot_main.cpp.

Ah right. Good catch. I'm still getting familiar with AOT :) 

> we could check if count() or sum() is even being used. IMO it'd be a little cleaner to have it there. Up to you.

I plan on not having this hack for too long so the less additional logic the better - besides are we really worried about the case where someone is running this on a machine with over 1024 cores?

Comment (danobi) [reply src/main.cpp]: > I plan on not having this hack for too long so the less additional logic the better - besides are we really worried about the case where someone is running this on a machine with over 1024 cores?

I think it's reasonable to be worried about this case. Looks like we are getting users working on supercomputers now

**Review by danobi — APPROVED** [review]
would be good to get another pair of eyes on this as well.

btw, does it make sense to document this behavior? maybe not if the original behavior was surprising. perhaps just a note about how it is expensive to read count() or sum() maps, as they are optimized for writing and not reading

**Comment by jordalgo** [comment]
> btw, does it make sense to document this behavior? 

Good call. I'll add something in the adoc and also make a changelog entry.

**Comment by danobi** [comment]
Had some shower thoughts:

1. Can simplify semantic analyser with the type helper: https://pastes.dxuuu.xyz/4ilo7i
1. I mentioned this above; not sure if we had different opinions or if I wasn't being clear, but the special case in the binop semantic analysis is a tad odd. For example, with this: https://pastes.dxuuu.xyz/a2w36w , you can still do `BEGIN { @ = count(); if ((int64)@ == 1) { print("yes") } }`. However, the cast (and thus potentially expensive operation) is explicit and provides some symmetry with how you synchronously access map-backed types in the rest of the language. 

I do also find it odd that for-looping over a map that contains `count()` as a value immediately collapses the map into an integer. This is probably wrong IMO. A user could be doing a loop and only wanting to do expensive stuff if the key is some value they want. Collapsing the value on an entry they don't immediately care about is bad for perf.

The advantage with the explicit casts is that we can change our minds later and make it implicit. Whereas if we start with implicit, we can't really take it back.

**Comment by jordalgo** [comment]
> Can simplify semantic analyser with the type helper

Good call. I'll update that.

> A user could be doing a loop and only wanting to do expensive stuff if the key is some value they care about. Collapsing the value on an entry they don't immediately care about is bad for perf. The advantage with the explicit casts is that we can change our minds later and make it implicit. Whereas if we start with implicit, we can't really take it back.

This is a good point. I also don't like the perf hit with implicit casting/collapsing - that being said I'm trying to think of use cases where a user is accessing a `count` or `sum` (or `min`, `max`, `avg`, which I plan on adding) value and just wants it for the current cpu. I feel like as long as the tradeoffs of using per-cpu map aggregations is documented (cheap writes, expensive reads) than the onus is on the user to choose the best map type for what they're doing with it. Also I feel like the auto-collapsing of a value inside a for-loop is better fixed by figuring out if the value is even being used and conditionally not collapsing it. In many ways I like when code is explicit but if we don't have good examples of using a `count` or `sum` value without collapsing then it just feels like cruft. I think this ties into your second point above but I need to think more on that.

**Comment by danobi** [comment]
> > Can simplify semantic analyser with the type helper
> 
> Good call. I'll update that.
> 
> > A user could be doing a loop and only wanting to do expensive stuff if the key is some value they care about. Collapsing the value on an entry they don't immediately care about is bad for perf. The advantage with the explicit casts is that we can change our minds later and make it implicit. Whereas if we start with implicit, we can't really take it back.
> 
> This is a good point. I also don't like the perf hit with implicit casting/collapsing - that being said I'm trying to think of use cases where a user is accessing a `count` or `sum` (or `min`, `max`, `avg`, which I plan on adding) value and just wants it for the current cpu. I feel like as long as the tradeoffs of using per-cpu map aggregations is documented (cheap writes, expensive reads) than the onus is on the user to choose the best map type for what they're doing with it. 

Yeah, I buy this argument. I can't think of any use cases where you'd not want the collapsed value. To take that idea to its logical conclusion, I think https://pastes.dxuuu.xyz/iydqri would work. Basically remove all the special cases and treat reading a count() as an integer. On write side it'll still be protected:

```
$ sudo ./build/src/bpftrace -e 'BEGIN { @x = count(); $xx = @x; $xx++; print($xx) }'
Attaching 1 probe...
2
^C

@x: 1
$ sudo ./build/src/bpftrace -e 'BEGIN { @x = count(); $xx = @x; $xx++; print($xx); @x = 5 }'
[sudo] password for dxu:
stdin:1:52-54: ERROR: Type mismatch for @x: trying to assign value of type 'int64' when map already contains a value of type 'count
BEGIN { @x = count(); $xx = @x; $xx++; print($xx); @x = 5 }

```

> Also I feel like the auto-collapsing of a value inside a for-loop is better fixed by figuring out if the value is even being used and conditionally not collapsing it. In many ways I like when code is explicit but if we don't have good examples of using a `count` or `sum` value without collapsing then it just feels like cruft. I think this ties into your second point above but I need to think more on that.

Presumably any loop that does filtering is on the slow or async path. So maybe this is good enough. Maybe in the long term once we enable user defined percpu maps then users could work around this if they really needed to.

File: man/adoc/bpftrace.adoc
Comment (danobi) [inline man/adoc/bpftrace.adoc:2423]:
```suggestion
Note: In contrast, multiple writers to a shared global var might lose counts as bpftrace 
doesn't update them atomically.
```

Comment (jordalgo) [reply man/adoc/bpftrace.adoc:2423]: Should it be "In contrast to `@++`, which uses a global hash map, multiple writers...

Comment (danobi) [reply man/adoc/bpftrace.adoc:2423]: Maybe add it on as "(eg `@++`)"

**Comment by danobi** [comment]
Oh, I think we need a feature probe in semantic analyser as well for the map elem callback helper

**Comment by jordalgo** [comment]
> Oh, I think we need a feature probe in semantic analyser as well for the map elem callback helper

Let me put up a separate PR for that.

**Comment by danobi** [comment]
Chatted offline about turning count() and sum() reads into proper ints. Type system does not allow us to do this easily. So I think all outstanding issues modulo feature probing are addressed

**Comment by jordalgo** [comment]
@ajor @viktormalik - Going to merge this soon unless you want more time to review.

File: src/ast/irbuilderbpf.cpp
Comment (viktormalik) [inline src/ast/irbuilderbpf.cpp:483]: This implements a non-trivial algorithm. Could we perhaps include a pseudo-code in the comment to make it clear what's the intention to do here?

File: src/ast/passes/semantic_analyser.cpp
Comment (viktormalik) [inline src/ast/passes/semantic_analyser.cpp:2338]: like this?
```suggestion
       !rhs.IsCtxAccess() && !rhs.IsArrayTy() && !rhs.IsCastableMapType()) ||
```

File: tests/codegen/per_cpu_map_cast.cpp
Comment (viktormalik) [inline tests/codegen/per_cpu_map_cast.cpp:37]: Missing EOL

**Review by viktormalik — APPROVED** [review]
LGTM, just a few nits

**Pull request merged**
